File: stack/02_min_stack/Main.java
import java.util.Stack;

public class Main {

    // Brute Force
    class BruteForce{
        private Stack<Integer> stack;
    
    
        public BruteForce(){
            stack = new Stack<>();
        }

        public void push(int val){
            stack.push(val);
        }
        public void pop(){
            stack.pop();
        }
        public int top(){
            return stack.peek();
        }
        public int getMin(){
            Stack<Integer> tmp = new Stack<>();
            int mini = stack.peek();

            while (!stack.isEmpty()){
                mini = Math.min(mini, stack.peek());
                tmp.push(stack.pop());
            }
            while(!tmp.isEmpty()){
                stack.push(tmp.pop());
            }
            return mini;
        }
    }


    class TwoStack{
        private Stack<Integer> stack;
        private Stack<Integer> minStack;

        public TwoStack(){
            stack = new Stack<>();
            minStack = new Stack<>();
        }
        public void push (int val){
            stack.push(val);
            if (minStack.isEmpty() || val <= minStack.peek()){
                minStack.push(val);
            }
        }
        public void pop(){
            if (stack.isEmpty()) return;
            int top = stack.pop();
            if (top == minStack.peek()){
                minStack.pop();
            }
        }
        public int top(){
            return stack.peek();
        }
        public int getMin(){
            return minStack.peek();
        }
    }
    class OneStack{
        long min;
        Stack<Long> stack;
        public OneStack(){
            stack = new Stack<>();
        }
        public void push(int x){
            if (stack.isEmpty()){
                stack.push(0L);
                min = x;
            }else{
                stack.push(x - min);
                if (x < 0)min = min - x;
            }
        }
        public void pop(){
            if (stack.isEmpty()) return;
            long pop = stack.pop();
            if (pop < 0) min = min - pop;
        }
        public int top(){
            long top = stack.peek();
            if (top > 0){
                return (int) (top + min);
            }else{
                return (int) min;
            }
        }
        public int getMin(){
            return (int) min;
        }    
    }

    

}

File: stack/02_min_stack/main.py


class BruteForce:
    def __init__(self):
        self.stack = []
    
    def push(self, val:int) -> None:
        self.stack.append(val)
    
    def pop(self) -> None:
        self.stack.pop()
    
    def top(self) -> int:
        return self.stack[-1]
    
    def getMin(self) -> int:
        tmp = []
        mini = self.stack[-1]
        while len(self.stack):
            mini = min(mini, self.stack[-1])
            tmp.append(self.stack.pop())
        while len(tmp):
            self.stack.append(tmp.pop())
        return mini
    

class TwoStack:
    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, val:int) -> None:
        self.stack.append(val)
        val = min(val, self.minStack[-1] if self.minStack else val)
        self.minStack.append(val)
    
    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()
    
    def top(self) -> int:
        return self.stack[-1]
    
    def getMin(self) -> int:
        return self.minStack[-1]


class OneStack:
    def __init__(self):
        self.min = float('inf')
        self.stack = []
    
    def push(self, x:int) -> None:
        if not self.stack:
            self.stack.append(0)
            self.min = x 
        else:
            self.stack.append(x - self.min)
            if x < self.min:
                self.min = x
    
    def pop(self) -> None:
        if not self.stack:
            return 
        pop = self.stack.pop()
        if pop < 0:
            self.min = self.min - pop
    
    def top(self) -> int:
        top = self.stack[-1]
        if top > 0:
            return top + self.min
        else:
            return self.min
    
    def getMin(self) -> int:
        return self.min